Unity-Technologies/Graphics
# Internal Inspector

Favro card: https://favro.com/organization/c564ede4ed3337f7b17986b6/d06b8de3e40da6d340b5b063?card=Uni-75466

---
### Purpose of this PR

The purpose of this PR is to introduce a new UI element called the Internal Inspector within the Shader Graph which will serve as the primary place for the user to interact with settings and data exposed on nodes and properties.

This window is contextual and reflects the currently selected node or property. 

The problem this is solving is to shift the settings away from the cramped cog wheel menu that is present on every node, which doesn't provide the ability to easily accommodate more settings in the future, and especially on HDRP presents a UX issue as the HDRP master nodes expose a ton of settings that are all squished into this tiny floating window:

![HDRP_StackLit](https://user-images.githubusercontent.com/13033542/79465556-55e0f800-7fb0-11ea-816f-6c81ecd76b84.PNG)

The Internal Inspector makes use of custom attributes and type inference (reflection) to extract properties from its targets, and then uses PropertyDrawers, quite similar to the API used to customize the Scene Inspector in the main editor window. (Link for API reference: https://docs.unity3d.com/ScriptReference/PropertyDrawer.html)

Things that are currently integrated into this system are :
- All Master nodes both HDRP and URP
- Blackboard properties (in code termed as BlackboardFieldViews)

Things that are NOT integrated into the system *yet*, (I want to get these done ASAP) are:
- PropertyNodeViews (i.e. property node pills in the graph itself, I know this is pretty critical so I think I'll address it in my next revision along with any bug fixes that are needed)
- CustomFunctionNode
- SubGraphOutputNode

---
### Testing status

**Manual Tests**: What did you do?
Added, edited the values of, renamed, duplicated and deleted all of the different property types from the Blackboard that don't require to be authored through code. Texture2DArray and Texture3D need to be tested particularly as I do not know how to author the assets needed to test those, some basic research I did revealed that they cannot be imported like simple textures for instance. 

Tested all the various master nodes to make sure I didnt leave any settings off the inspector, also made sure to change the various values on them and make sure they respond similarly to how they did before, there are a lot of conditional things happening in the settings of the master nodes based on the values of other settings so thats a prime location for bugs.

Will add more gifs and details to round out the PR.
